tests/examples/shims/root.lua

```diff
@@ -1,9 +1,126 @@
 local root = {_tags={}}
 
+local hotkeys = nil
+
 function root:tags()
     return root._tags
 end
 
+function root:size() --TODO use the screens
+    return 0, 0
+end
+
+function root:size_mm()
+    return 0, 0
+end
+
+function root.cursor() end
+
+-- GLOBAL KEYBINDINGS --
+
+local keys = {}
+
+-- Note: the tests are not expected to create 1000 of them, so to keep the
+-- implementation minimal, it doesn't try to hash them and simply loop over
+-- every entry.
+function root.keys(k)
+    keys = k or keys
+    return keys
+end
+
+-- FAKE INPUTS --
+
+local conversion, mods = {
+    Super_L   = "Mod4",
+    Control_L = "Control",
+    Shift_L   = "Shift",
+    Alt_L     = "Mod1",
+    Super_R   = "Mod4",
+    Control_R = "Control",
+    Shift_R   = "Shift",
+    Alt_R     = "Mod1",
+}, {}
+
+local function add_modkey(key)
+    if not conversion[key] then return end
+    mods[conversion[key]] = true
+end
+
+local function get_mods()
+    local ret = {}
+
+    for mod in pairs(mods) do
+        table.insert(ret, mod)
+    end
+
+    return ret
+end
+
+local function remove_modkey(key)
+    if not conversion[key] then return end
+    mods[conversion[key]] = nil
+end
```

Comment:
I suggest moving `add_modkey` and `remove_modkey` next to each other. They belong together.

Also, technically these functions are wrong. Which modifiers are active when I
- press `Alt_L`
- press `Alt_R`
- relese `Alt_L`.

Your code says that no modifiers are active, my intuition tells me that `Mod1` should be active.

Reply:
I know, but the goal of the shims is to be the most minimal implementation that pass the currently existing tests. It was never intended to behave correctly and handle corner cases that are not reached using the documentation tests.

Adding all these corner cases (and also the keycode vs. keysym equivalent of this bug) would make the code harder to read for no practical reasons.